FuelLabs/fuel-indexer
# enhancement: generate GraphQL schema from JSON ABI

### Description

Closes #846.

This PR adds a `--json-abi` flag to `forc index new` which facilitates automatic generation of the GraphQL schema from the given JSON ABI.

### Testing steps

Create a new indexer:

```
cargo run -p forc-index -- new --json-abi ./packages/fuel-indexer-tests/trybuild/abi/DAO-contract-abi.json dao-indexer
```

Copy the JSON ABI:

```
cp ./packages/fuel-indexer-tests/trybuild/abi/DAO-contract-abi.json /tmp/dao-indexer
```

Set the `abi` in the `Manifest`

```
abi: DAO-contract-abi.json
```

Change `lib.rs` to refer to types from the generated schema:

```
extern crate alloc;
use fuel_indexer_utils::prelude::*;

#[indexer(manifest = "dao_indexer.manifest.yaml")]
pub mod dao_indexer_index_mod {

    fn dao_indexer_handler(ie: InitializeEvent) {
        info!("author: {:?}", ie.author);
        let iee = InitializeEventEntity::new(ie.author, ie.token);
        iee.get_or_create();
    }
}
```

Verify that it builds:

```
forc index build
```

### Changelog

* Add `generate_schema()`, which generates a GraphQL schema from JSON ABI.
* Add documentation.
* Move some functions from `packages/fuel-indexer-macros/src/helpers.rs` to `packages/fuel-indexer-lib/src/helpers.rs` because they are now used in both places.
